### PR TITLE
move EC_METHOD to internal-only

### PR DESCRIPTION
**Pull request by bbbrumley**
Last TODO item from #8332! Quoting from there:

* Entire concept of `EC_METHOD` should not be public facing.

It was surprisingly easy to update the tests.

~~It's WiP because I'll document everything when I get some high level confirmation that:~~

1. ~~It's OK to deprecate all these functions~~
2. ~~It's OK to add `EC_GROUP_get_field_type`~~

Edit: finally out of WiP after ea1cd6bfbce0a10102a275fa73536c91896e990f. Thanks @levitte @t8m and @romen for the feedback.

### CODE REVIEW

**Comment by levitte** [comment]
This looks good to me.

**Comment by bbbrumley** [comment]
I think the travis red cross is relevant, and to be honest when I submitted the PR I had no idea how this was actually working at all since I marked those public `EC_METHOD`s as deprecated. Yet both

1. `./config -d no-shared no-deprecated -Wall -Werror -Wl,--fatal-warnings`
2. `./config -d no-shared -Wall -Werror -Wl,--fatal-warnings`

succeeded, so I blissfully ignored.

But now I think I see the trick going on internally -- is this the right fix for the red cross?

```diff
diff --git a/crypto/ec/ecp_s390x_nistp.c b/crypto/ec/ecp_s390x_nistp.c
index 75c8475e69..edbad15cdd 100644
--- a/crypto/ec/ecp_s390x_nistp.c
+++ b/crypto/ec/ecp_s390x_nistp.c
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * EC_METHOD low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
+
 #include <stdlib.h>
 #include <string.h>
 #include <openssl/err.h>
```

**Comment by levitte** [comment]
Yes, that's the trick.

**Comment by bbbrumley** [comment]
> Yes, that's the trick.

Thanks @levitte! All green after 0bf4176966b0463a4c7580940f38b6501d93e4ba.

File: crypto/ec/ec_lib.c [hunk 1]
```diff
@@ -475,6 +479,11 @@ const BIGNUM *EC_GROUP_get0_field(const EC_GROUP *group)
     return group->field;
 }
 
+int EC_GROUP_get_field_type(const EC_GROUP *group)
+{
+    return group->meth->field_type;
```
Comment (romen) [inline crypto/ec/ec_lib.c:484]: Related: I think it is fine here not to check if group->meth is NULL before dereferencing it.

File: test/ectest.c [hunk 2]
```diff
@@ -836,8 +802,7 @@ static int char2_curve_test(int n)
         || !TEST_true(BN_hex2bn(&p, test->p))
         || !TEST_true(BN_hex2bn(&a, test->a))
         || !TEST_true(BN_hex2bn(&b, test->b))
-        || !TEST_true(group = EC_GROUP_new(EC_GF2m_simple_method()))
-        || !TEST_true(EC_GROUP_set_curve(group, p, a, b, ctx))
```
Comment (romen) [inline test/ectest.c:840]: Sidenote: do you have thoughts about deprecating also EC_GROUP_set_curve? (as another PR in case it makes sense)

File: test/ectest.c [hunk 3]
```diff
@@ -2170,7 +2119,7 @@ static int cardinality_test(int n)
 
     if (!TEST_ptr(ctx = BN_CTX_new())
         || !TEST_ptr(g1 = EC_GROUP_new_by_curve_name(nid))
-        || !TEST_ptr(g2 = EC_GROUP_new(EC_GROUP_method_of(g1)))) {
+        || !TEST_ptr(g2 = EC_GROUP_new_by_curve_name(nid))) {
```
Comment (romen) [inline test/ectest.c:2122]: I guess this test could be just dropped now?
I think it was explicitly testing EC_GROUP_method_of, now is not doing much really!

File: test/ectest.c (hunk 2 above)
Comment (romen) [reply test/ectest.c:840]: To elaborate more:

Once `EC_GROUP_new(some_method)` function is deprecated, I don't see leftover reasons to set parameters after instantiation.

File: crypto/ec/ec_curve.c [hunk 4]
```diff
@@ -3388,17 +3388,15 @@ int ec_curve_nid_from_params(const EC_GROUP *group, BN_CTX *ctx)
...
     const EC_CURVE_DATA *data;
     const EC_POINT *generator = NULL;
-    const EC_METHOD *meth;
     const BIGNUM *cofactor = NULL;
     /* An array of BIGNUMs for (p, a, b, x, y, order) */
     BIGNUM *bn[NUM_BN_FIELDS] = {NULL, NULL, NULL, NULL, NULL, NULL};
 
-    meth = EC_GROUP_method_of(group);
-    if (meth == NULL)
+    if (group->meth == NULL)
```
[the remark this answers is not in the document]
Comment (bbbrumley) [reply crypto/ec/ec_curve.c:3395]: I agree. It's just old habits die hard and it was quick edits preserving what was there :) But I'll update it

File: test/ectest.c (hunk 3 above)
Comment (bbbrumley) [reply test/ectest.c:2122]: I ... think I wrote that line. Further down in the function, it's manipulating `g2` to muck with the parameters, then recalculating them and checking vs `g1`. So `g1` is kind of the "control" `EC_GROUP` for the test.

Comment (romen) [reply test/ectest.c:2122]: Oh, yeah, I missed how g2 was being used!

File: test/ectest.c (hunk 2 above)
Comment (bbbrumley) [reply test/ectest.c:840]: I agree in the sense that ... the only time I've written `EC_GROUP_set_curve` in code is exactly the file you're looking at. To be able to reuse an `EC_GROUP` without re-initializing it.

But outside of a test, I don't see why anyone would do that.

That is ofc just my subjective opinion -- I have not checked for applications using `EC_GROUP_set_curve`.

File: test/ectest.c (hunk 3 above)
Comment (bbbrumley) [reply test/ectest.c:2122]: But yea you're right I was lazy there -- it would be better to not instantiate `g1` and `g2` literally from the same call, to keep them as independent as possible for a valid control. I'll adjust it

**Comment by bbbrumley** [comment]
Right so latest red cross is because I switched `nistp_single_test` from a method-specific test per NIST GFp curve to a curve NID-specific test per NIST GFp curve. And the `ecp_s390x_nistp.c` `EC_METHOD`s do not support precomp. I'll push a fixup.

File: test/ectest.c (hunk 2 above)
Comment (romen) [reply test/ectest.c:840]: Maybe we can add this item to the parent issue you were using to track all these related PRs?
We can try to do an analysis of that and see if deprecation can be recommended.

File: test/ectest.c [hunk 5]
```diff
@@ -1396,7 +1344,7 @@ static int nistp_single_test(int idx)
         || !TEST_false(EC_GROUP_have_precompute_mult(NISTP))
         /* now repeat all tests with precomputation */
         || !TEST_true(EC_GROUP_precompute_mult(NISTP, ctx))
-        || !TEST_true(EC_GROUP_have_precompute_mult(NISTP))
+        || !TEST_int_ge(EC_GROUP_have_precompute_mult(NISTP), 0)
```
Comment (romen) [inline test/ectest.c:1347]: We can effectively strip this, right?
We are testing that the value is 0 or greater for a function that should return either 0 or 1.

File: test/ectest.c [hunk 6]
```diff
@@ -1430,7 +1378,7 @@ static int nistp_single_test(int idx)
         || !TEST_int_eq(0, EC_POINT_cmp(NISTP, Q, G, ctx)))
       goto err;
 
-    r = group_order_tests(NISTP);
+    r = 1;
```
Comment (romen) [inline test/ectest.c:1381]: We are removing this, because it is effectively already tests in `prime_field_tests()` and this now doesn't add any value anymore, right?

Comment (bbbrumley) [reply test/ectest.c:1381]: Yes exactly

File: test/ectest.c (hunk 2 above)
Comment (bbbrumley) [reply test/ectest.c:840]: I added `EC_GROUP_set_curve` to the #8332 list for analysis.

File: test/ectest.c (hunk 5 above)
Comment (bbbrumley) [reply test/ectest.c]: done in ab7aa90af9c1ceb4c8df71d33c4b4e5805efd6e2

**Comment by bbbrumley** [comment]
> A few more comments

Thanks. I will start on the `pod` and `CHANGES.md` entries hopefully today.

**Comment by romen** [comment]
Thanks for all the work so far!

File: crypto/ec/ec_lib.c [hunk 7]
```diff
@@ -23,6 +23,7 @@
 
 /* functions for EC_GROUP objects */
 
+#ifndef OPENSSL_NO_DEPRECATED_3_0
 EC_GROUP *EC_GROUP_new_ex(OPENSSL_CTX *libctx, const EC_METHOD *meth)
 {
```
Comment (t8m) [inline crypto/ec/ec_lib.c:28]: It does not make sense to make a new 3.0 function immediately deprecated. Should it be added at all?

File: include/openssl/ec.h [hunk 8]
```diff
@@ -113,15 +115,15 @@ const EC_METHOD *EC_GF2m_simple_method(void);
  *  \param   meth   EC_METHOD to use
  *  \return  newly created EC_GROUP object or NULL in case of an error.
  */
-EC_GROUP *EC_GROUP_new_ex(OPENSSL_CTX *libctx, const EC_METHOD *meth);
+DEPRECATEDIN_3_0(EC_GROUP *EC_GROUP_new_ex(OPENSSL_CTX *libctx, const EC_METHOD *meth))
```
Comment (t8m) [inline include/openssl/ec.h:118]: As said above - it does not make sense to immediately deprecate a new function so it should be converted to an internal function instead.

File: crypto/ec/ec_lib.c (hunk 7 above)
Comment (romen) [reply crypto/ec/ec_lib.c:28]: Good point!

@bbrumley I believe this will also affect your documentation edits!

File: include/openssl/ec.h (hunk 8 above)
Comment (bbbrumley) [reply include/openssl/ec.h:118]: tyvm @t8m ! I will now sanity check all of these, since that never even crossed my mind ...

Comment (bbbrumley) [reply include/openssl/ec.h]: > it does not make sense to immediately deprecate a new function so it should be converted to an internal function instead

Done in 17468cdf115b92d47ecb29952ec3dbd14d5b7ebd. I had to rebase due to conflict but I didn't squash yet just to keep the progress.

I sanity checked the rest of the functions for good measure. They're all ancient, so this was the only one.

Thanks for catching it :)

File: util/libcrypto.num [hunk 9]
```diff
@@ -4682,7 +4682,7 @@ ERR_set_error                           ?	3_0_0	EXIST::FUNCTION:
 ERR_vset_error                          ?	3_0_0	EXIST::FUNCTION:
 X509_get0_authority_issuer              ?	3_0_0	EXIST::FUNCTION:
 X509_get0_authority_serial              ?	3_0_0	EXIST::FUNCTION:
-EC_GROUP_new_ex                         ?	3_0_0	EXIST::FUNCTION:DEPRECATEDIN_3_0,EC
+EC_GROUP_new_ex                         ?	3_0_0	NOEXIST::FUNCTION:DEPRECATEDIN_3_0,EC
```
Comment (bbbrumley) [inline util/libcrypto.num:4685]: @t8m well above is `make update` result but I don't think it really makes sense?

Comment (t8m) [reply util/libcrypto.num:4685]: If you remove manually the line and re-run make update, it should not reappear.

File: util/libcrypto.num [hunk 10]
```diff
@@ -5094,4 +5093,6 @@ EVP_PKEY_CTX_set_rsa_pss_keygen_saltlen ?	3_0_0	EXIST::FUNCTION:RSA
 EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md ?	3_0_0	EXIST::FUNCTION:RSA
 EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md_name ?	3_0_0	EXIST::FUNCTION:RSA
 OSSL_PROVIDER_do_all                    ?	3_0_0	EXIST::FUNCTION:
+EC_GROUP_get_field_type                 ?	3_0_0	EXIST::FUNCTION:EC
+OSSL_CRMF_CERTID_dup                    ?	3_0_0	EXIST::FUNCTION:CRMF
```
Comment (t8m) [inline util/libcrypto.num:5097]: Apparently in some other PR a make update commit was missing? These additions should go into a separate commit.

Comment (bbbrumley) [reply util/libcrypto.num:5097]: It's just an artifact of the rebase. It'll fall out when I squash stuff. (I also noted and looked into the same before I pushed.)

File: util/libcrypto.num (hunk 9 above)
Comment (bbbrumley) [reply util/libcrypto.num:4685]: perfect ty that works.

I'll do the `pod` and `CHANGES.md` edits now, then squash and this one should be ready.

**Comment by bbbrumley** [comment]
So I added the doc and squashed. Renamed to take out of WiP. Let me know if there's anything else.

**Comment by t8m** [comment]
Would you please rebase to fix the CHANGES.md conflict?

**Comment by bbbrumley** [comment]
> Would you please rebase to fix the CHANGES.md conflict?

Done :)

**Comment by romen** [comment]
Merging this now!

**Comment by romen** [comment]
Merge to `master` as 23ccae80bd58adfe89e3e345414684eb82bdb531

Thanks all!
